api/include/opentelemetry/trace/propagation/b3_propagator.h

```diff
@@ -0,0 +1,276 @@
+// headers of HTTP requests. HTTP frameworks and clients can integrate with B3Propagator by
+// providing the object  containing the headers, and a getter function for the extraction.
+// Based on:
+// https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#b3-extract
+template <typename T>
+class B3PropagatorExtractor : public HTTPTextFormat<T>
+{
+public:
+  // Rules that manages how context will be extracted from carrier.
+  using Getter = nostd::string_view (*)(const T &carrier, nostd::string_view trace_type);
+
+  // Returns the context that is stored in the HTTP header carrier with the getter as extractor.
+  context::Context Extract(Getter getter,
+                           const T &carrier,
+                           context::Context &context) noexcept override
+  {
+    SpanContext span_context = ExtractImpl(getter, carrier);
+    nostd::shared_ptr<Span> sp{new DefaultSpan(span_context)};
+    return context.SetValue(kSpanKey, sp);
+  }
+
+  static SpanContext GetCurrentSpan(const context::Context &context)
+  {
+    context::Context ctx(context);
+    context::ContextValue span = ctx.GetValue(kSpanKey);
+    if (nostd::holds_alternative<nostd::shared_ptr<Span>>(span))
+    {
+      return nostd::get<nostd::shared_ptr<Span>>(span).get()->GetContext();
+    }
+    return SpanContext::GetInvalid();
+  }
+
+  static TraceId GenerateTraceIdFromString(nostd::string_view trace_id)
+  {
+    uint8_t buf[kTraceIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateBuffFromHexStrPad0(trace_id, sizeof(buf), buf);
+    return TraceId(buf);
+  }
+
+  static SpanId GenerateSpanIdFromString(nostd::string_view span_id)
+  {
+    uint8_t buf[kSpanIdHexStrLength / 2];
+    uint8_t *b_ptr = buf;
+    GenerateBuffFromHexStrPad0(span_id, sizeof(buf), buf);
+    return SpanId(buf);
+  }
+
+  static TraceFlags GenerateTraceFlagsFromString(nostd::string_view trace_flags)
+  {
+    if (trace_flags.length() != 1 || (trace_flags[0] != '1' && trace_flags[0] != 'd'))
+    {  // check for invalid length of flags and treat 'd' as sampled
+      return TraceFlags(0);
+    }
+    return TraceFlags(TraceFlags::kIsSampled);
+  }
+
+private:
+  // Converts hex numbers (string_view) into bytes stored in a buffer and pads buffer with 0.
+  static void GenerateBuffFromHexStrPad0(nostd::string_view hexStr, int bufSize, uint8_t *buf)
+  {  // we are doing this starting from "right" side for left-padding
+    nostd::string_view::size_type posInp = hexStr.length();
+    int posOut                           = bufSize;
+    while (posOut--)
+    {
+      int val = 0;
+      if (posInp)
+      {
+        int hexDigit2 = HexToInt(hexStr[--posInp]);  // low nibble
+        int hexDigit1 = 0;
+        if (posInp)
+        {
+          hexDigit1 = HexToInt(hexStr[--posInp]);
+        }
+        if (hexDigit1 < 0 || hexDigit2 < 0)
```

Comment:
This branch is never taken as `HexToInt` returns `uint8_t` (`uint8_t(-1) = int(255)`), thus a malformed hex sequence is generated.

Reply:
Yep :( copied from here... https://github.com/open-telemetry/opentelemetry-cpp/blob/ce89aa10941c14a6e52e66e68e5c1831c143eaf2/api/include/opentelemetry/trace/propagation/http_trace_context.h#L120-L123